### DISCUSSION TITLE
Custom volar plugins stop working in v1.6

### DISCUSSION THREAD

**Discussion by kingyue737** [vuejs/language-tools#2752] [Q&A] [answered]
I've turned on log of Vue semantic and syntactic in extension. Then I added many `console.log('...')` in my plugin code and in `volar.config.js`. But none of them are printed in terminal. It seemed that `volar.config.js` is not loaded. Is there a way to see the log or debug my own plugin?

**Comment by johnsoncodehk** [comment]
Have you update the `@volar-plugins/xxx` module to v2.0.0 later? We changed the export method, now the plugin is exported with `export default`, so you need to change the previous `require('@volar-plugins/xxx')` to `require('@volar-plugins/xxx').default`.

You can try https://github.com/johnsoncodehk/volar-starter, if it still doesn't work, please provide a repro case.

**Reply by johnsoncodehk** [reply]
Please note that we have also changed the plugin API design.

https://github.com/volarjs/volar.js/issues/14

**Reply by kingyue737** [reply]
I tried 3 repos. But all failed. 
-  Repo A with my custom plugin. I've already updated my plugin according to new API, but nothing happened.
- Repo B with `volar-service-vetur`. I updated `volar-service-vetur` or `@volar-plugins/vetur` to latest, but no intellisense or snippets triggered.
- `volar-starter`:
    1. Clone the repo, then `pnpm install`
    2. Update `volar.config.js` with `require("@volar-plugins/prettyhtml").default({tabWidth: 8})`
    3. Restart Volar Server
    4. Open `App.vue`. Open command palette. `Format document with Vue Language Feature Volar (Volar)`
    5. Tab width is not changed. 

It just seemed like that `volar.config.js` does not work at all.🤔

**Reply by johnsoncodehk** [reply]
Can you try this to have an effect in volar-starter?
```js
// volar.config.js
module.exports = {
    plugins: [
        () => ({
            provideDocumentFormattingEdits(document) {
                return [{
                    newText: 'hello world',
                    range: {
                        start: document.positionAt(0),
                        end: document.positionAt(document.getText().length),
                    },
                }];
            },
        }),
    ],
};
```

**Reply by kingyue737** [reply]
Does this plugin expect to rewrite the file with "hello world" when formatting?

Updated to v1.6.4. Restart Volar server and format a document. Nothing happened.

**Reply by johnsoncodehk** [reply]
Can you post your VSCode settings?

**Reply by kingyue737** [reply]
`volar-starter/.vscode/settings.json` does not exist.

I haven't changed any setting of volar. The only field related in `User/settings.json` may be `[vue].editor.defaultFormatter`, but I did format 
 document with volar via command palette. Maybe a platform related issue on Windows?

User/settings:
```json  
"terminal.integrated.inheritEnv": false,
  "git.enableSmartCommit": true,
  "git.autofetch": true,
  "vim.foldfix": true,
  "launch": {
    "version": "0.2.0",
    "configurations": [
      {
        "name": "Python: Global",
        "type": "python",
        "request": "launch",
        "program": "${file}",
        "console": "integratedTerminal"
      }
    ],
    "compounds": []
  },
  "python.testing.pytestEnabled": true,
  "editor.formatOnPaste": true,
  "vim.handleKeys": {
    "<C-c>": false,
    "<C-x>": false,
    "<C-y>": false,
    "<C-f>": false,
    "<C-d>": true
  },
  "settingsSync.ignoredSettings": [],
  "jupyter.askForKernelRestart": false,
  "explorer.confirmDelete": false,
  "jupyter.debugJustMyCode": false,
  "jupyter.stopOnFirstLineWhileDebugging": false,
  "workbench.editorAssociations": {
    "*.ipynb": "jupyter-notebook"
  },
  "terminal.integrated.profiles.windows": {
    "PowerShell": {
      "source": "PowerShell",
      "icon": "terminal-powershell"
    },
    "Command Prompt": {
      "path": [
        "${env:windir}\\Sysnative\\cmd.exe",
        "${env:windir}\\System32\\cmd.exe"
      ],
      "args": [],
      "icon": "terminal-cmd"
    },
    "Git Bash": {
      "source": "Git Bash",
      "path": ["D:\\Program Files\\Git\\bin\\bash.exe"],
      "icon": "terminal-bash"
    }
  },
  "terminal.integrated.defaultProfile.windows": "PowerShell",
  "terminal.external.windowsExec": "%userprofile%\\\\AppData\\\\Local\\\\Microsoft\\\\WindowsApps\\\\wt.exe",
  "terminal.integrated.env.windows": {
    "http_proxy": "http://127.0.0.1:7890",
    "https_proxy": "http://127.0.0.1:7890"
  },
  "security.workspace.trust.untrustedFiles": "open",
  "[vue]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[typescript]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[jsonc]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[javascript]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "terminal.integrated.tabs.focusMode": "singleClick",
  "diffEditor.ignoreTrimWhitespace": false,
  "editor.codeActionsOnSave": {
    "source.fixAll.eslint": true
  },
  "svg.preview.mode": "svg",
  "[json]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "notebook.cellToolbarLocation": {
    "default": "right",
    "jupyter-notebook": "left"
  },
  "notebook.output.textLineLimit": 3000,
  "[html]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "i18n-ally.displayLanguage": "zh",
  "[markdown]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "vite.autoStart": false,
  "vite.buildCommand": "pnpm run build",
  "vite.devCommand": "pnpm run dev",
  "vite.port": 9527,
  "json.schemas": [
    {
      "fileMatch": ["/.prettierrc"],
      "url": "https://json.schemastore.org/prettierrc.json"
    }
  ],
  "explorer.confirmDragAndDrop": false,
  "python.linting.mypyEnabled": true,
  "[scss]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "terminal.integrated.fontFamily": "Consolas, 'Courier New', monospace, CaskaydiaCove NF Mono",
  "totalTypeScript.hideBasicTips": true,
  "totalTypeScript.hideAllTips": true,
  "editor.formatOnSave": true,
  "[css]": {
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
  "[python]": {
    "editor.formatOnType": true
  },
  "jupyter.interactiveWindow.creationMode": "perFile"
}
```

**Reply by johnsoncodehk** [reply]
You're right, please track https://github.com/volarjs/volar.js/issues/36.

Unfortunately this won't be fixed until the v1.7 pre-release at the earliest.

**Reply by hunyan-io** [reply, marked as answer]
Hi, I can see you're on windows. I also have the same issue. This happens because of https://github.com/volarjs/volar.js/issues/36
Another thing which I find a bit weird is that all examples including the one here use the `plugins` option, but reading the source code I believe it should be `services`. Fixing the issue I linked and switching to `services` fixes the issue for me.

**Reply by johnsoncodehk** [reply]
<= v1.6.x (v1 branch of this repo, based on Volar.js v1.4) using "plugins".
\>= v1.7 pre-release ~ v2.0 (master branch of this repo, based on Volar.js v1.5) using "services".

For v1.6.4 (the last stable release of v1), you should use "plugins"
